### DISCUSSION TITLE
I can't get the document errors correctly

### DISCUSSION THREAD

**Discussion by kjlsdlkfio332fsdkla32skjdf** [pymupdf/PyMuPDF#4021] [Looking for help] [answered]
### Description of the bug

I use the TOOLS code to get a warning when I open a document. The fact is that I need to get all the structural errors of the document in this way (xref table, incorrect stream), etc. But for some reason, errors don't always work correctly. Please tell me, can I somehow get all the structural errors that it contains from the document class? or what solution should I use for this?
![image](https://github.com/user-attachments/assets/40303283-94ee-4dc7-81af-e8f6971e1252)
![image](https://github.com/user-attachments/assets/442caefc-c9cb-457c-9b75-876b54de5ce6)
how can I conveniently get errors for specific files?

### How to reproduce the bug

When I try to get errors in this way from several files at once, simultaneously clearing the buffer - it sometimes accumulates, and sometimes even works incorrectly (without displaying a list of errors in principle). 

def check_struct(file_path):
    pymupdf.TOOLS.mupdf_display_warnings(False)
    pymupdf.TOOLS.mupdf_display_errors(False)
    document = pymupdf.open(file_path)
    document.close()
    result = pymupdf.TOOLS.mupdf_warnings(True)
    return result

print(check_struct("1.pdf"))

### PyMuPDF version

1.24.13

### Operating system

Windows

### Python version

3.11

**Comment by JorjMcKie** [comment]
This is a typical Discussions item - transferring ...

**Reply by kjlsdlkfio332fsdkla32skjdf** [reply, marked as answer]
and what i can do ?

**Reply by kjlsdlkfio332fsdkla32skjdf** [reply]
I can't get a normal response from the developers in any way.

**Comment by JorjMcKie** [comment]
> Please tell me, can I somehow get all the structural errors that it contains from the document class? or what solution should I use for this?

I tried to answer this over on Discord already:
Without examining every bit and piece, there never can be a guarantee that a PDF is error-free!
The PDF file format contains no central information about its integrity. While many Cross Reference Table problems are detected during open (but again **_not all of them!_**), even the correctness of the page tree will not be integrity-checked at that time.
* To check the page tree integrity you may have to load all pages - a **_terrible waste of resources_** considering that errors here have a probability far below 1%.
* Even if the page tree is okay, pages may contains errors that only show when you extract text, or only when you render them, or ....
* If the PDF uses `/ObjStm` objects you may have to try and read all xrefs to confirm that the respective object exists (may easily be many thousands of them!). And at that point you still haven't checked that that object is error-free and represents the expected type.
* A page may contain illegal items in its `/Annots` array. This cannot be detected until you load all annotations from all pages. Annotations too are an object type that has content streams.

I could endlessly go on like this. PDF's essentially are ASCII text files with interspersed binary parts. Just simply anything can be wrong with them and most of the errors are not detectable at open time.

**Reply by JorjMcKie** [reply]
BTW:
(Py-) MuPDF uses the same reasonable, economic approach as most other tools. Have you never experienced a case where Adobe Acrobat reader displayed all pages of a PDF ... except one - which was only detected when you actually tried it?

**Reply by kjlsdlkfio332fsdkla32skjdf** [reply]
I'm trying to describe another problem. Read all or part of the contents - I'll figure it out. I need to somehow get the errors that the library gives out in the right format. I gave an example. I am not satisfied with the output to the console or logger. Why can't I make it so that I can catch the error using Try Except?

**Reply by kjlsdlkfio332fsdkla32skjdf** [reply]
Or will an exception be thrown if I try to read the contents of the pages? (what I have depicted in the screenshot)

**Reply by JorjMcKie** [reply]
Exceptions are only thrown if there is a must ... no way out for any reasonable processing.
This decision is made on the MuPDF level for the overwhelming part.
MuPDF's priority mindset is to enable processing where possible.
But you are also free to interpret the string output in `TOOLS.mupdf_warnings()`.

If your type of processing requires chemically clean PDFs (for reasons beyond my imagination) you should probably look for specialized tools. Ghostscript may be a good start `gs -dNOPAUSE -dBATCH -sDEVICE=nullpage yourfile.pdf`.
I repeat:
For economical reasons I would always recommend to assume that a PDF is in good shape (= optimistic approach) and copy problem cases to an error folder for later clearance.

If you do not deal with notoriously dirty file sources, this should yield the best overall result and throughput.
